File: PartitionList.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @param x, an integer
    # @return a ListNode
    # http://www.cnblogs.com/zuoyuan/p/3783276.html
    def partition(self, head, x):
        lhead = ListNode(0)
        ghead = ListNode(0)
        tmp = head
        pl, pg = lhead, ghead
        while tmp:
            if tmp.val < x:
                pl.next = tmp
                tmp = tmp.next
                pl = pl.next
                pl.next = None
            else:
                pg.next = tmp
                tmp = tmp.next
                pg = pg.next
                pg.next = None
        pl.next = ghead.next
        head = lhead.next
        return head
        
        

File: CPP/Second/partition_list.cc
// https://oj.leetcode.com/problems/partition-list/

// Given a linked list and a value x, partition it such that all nodes less than x
// You should preserve the original relative order of the nodes in each of the two
// For example,
// Given 1->4->3->2->5->2 and x = 3,
// return 1->2->2->4->3->5.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    ListNode *partition(ListNode *head, int x) {
        ListNode dummy1(-1);
        ListNode dummy2(-1);
        ListNode *p = &dummy1, *q = &dummy2;
        for (ListNode *t = head; t; t = t->next) {
            if (t->val < x) {
                p->next = new ListNode(t->val);
                p = p->next;
            }
            else {
                q->next = new ListNode(t->val);
                q = q->next;
            }
        }
        p->next = dummy2.next;
        q->next = nullptr;
        return dummy1.next;
    }
};
int main(int argc, char *argv[])
{

    return 0;
}
